### PR TITLE
Support null token in spnego

### PR DESCRIPTION
**Pull request by pschorf**

### CODE REVIEW

**Comment by shamsmahmood** [comment]
lgtm

File: scheduler/src/cook/spnego.clj [hunk 1]
```diff
@@ -78,13 +78,14 @@
   [rh]
   (fn [req]
     (if (get-in req [:headers "authorization"])
-      (let [gss_context (gss-context-init)]
-        (if-let [token (do-gss-auth-check gss_context req)]
+      (let [gss_context (gss-context-init)
+            token (do-gss-auth-check gss_context req)]
+        (if (.isEstablished gss_context)
```
Comment (wyegelwel) [inline scheduler/src/cook/spnego.clj]: Can you add a comment what is going on here?

File: scheduler/src/cook/spnego.clj [hunk 2]
```diff
@@ -78,13 +78,14 @@
...
-            (-> req
-                (assoc :krb5-authenticated-princ princ
-                       :authorization/user (first (str/split princ #"@" 2)))
-                (rh)
-                (header "WWW-Authenticate" token)))
+            (cond-> (-> req
+                        (assoc :krb5-authenticated-princ princ
+                               :authorization/user (first (str/split princ #"@" 2)))
+                        (rh))
+              token (header "WWW-Authenticate" token)))
```
Comment (wyegelwel) [inline scheduler/src/cook/spnego.clj]: What is the implication of not including this WWW-Authenticate when the token doesn't exist?

Comment (pschorf) [inline scheduler/src/cook/spnego.clj]: As I understand the spec, SPNEGO can have an arbitrary number of back and forth challenges. The way it's used in Cook right now just expects a challenge from the client, then a response from the server authenticating the server.

It's possible for a client not to request the server to authenticate itself, in which case the token is nil and there's no need to send a WWW-Authenticate header back to the client: just treat the session as authenticated (in Java, you can accomplish this by calling GSSContext.requestMutualAuth(false) on the client GSSContext before making the request.)

File: scheduler/src/cook/spnego.clj (hunk 1 above)
Comment (pschorf) [inline scheduler/src/cook/spnego.clj]: Added a comment

**Comment by dposada** [comment]
LGTM, can we merge?

**Comment by pschorf** [comment]
Yes I think it should be good.

**Pull request merged**